File: Tests/DATAObjectIDs/DATAObjectIDsTests.m
@import XCTest;

#import "Tests-Swift.h"
#import "User+CoreDataClass.h"
@import DATAStack;

@interface DATAObjectIDsTests : XCTestCase

@end

@implementation DATAObjectIDsTests

- (User *)insertUserWithRemoteID:(NSNumber *)remoteID
                         localID:(NSString *)localID
                            name:(NSString *)name
                       inContext:(NSManagedObjectContext *)context {
    User *user = [NSEntityDescription insertNewObjectForEntityForName:@"User"
                                               inManagedObjectContext:context];
    user.remoteID = remoteID;
    user.localID = localID;
    user.name = name;

    return user;
}
- (void)configureUserWithRemoteID:(NSNumber *)remoteID
                          localID:(NSString *)localID
                             name:(NSString *)name
                            block:(void (^)(User *user, NSManagedObjectContext *context))block {
    DATAStack *stack = [[DATAStack alloc] initWithModelName:@"Tests"
                                                     bundle:[NSBundle bundleForClass:[self class]]
                                                  storeType:DATAStackStoreTypeInMemory];

    [stack performInNewBackgroundContext:^(NSManagedObjectContext *context) {
        User *user = [self insertUserWithRemoteID:remoteID localID:localID name:name inContext:context];

        NSError *error = nil;
        if (![context save:&error]) {
            NSLog(@"error saving: %@", error);
            abort();
        }

        if (block) {
            block (user, context);
        }
    }];

}

- (void)testDictionary {
    [self configureUserWithRemoteID:@1 localID:nil name:@"Joshua" block:^(User *user, NSManagedObjectContext *context) {
        NSDictionary *dictionary = [DATAObjectIDs objectIDsInEntityNamed:@"User"
                                                     withAttributesNamed:@"remoteID"
                                                                 context:context];
        XCTAssertNotNil(dictionary);
        XCTAssertTrue(dictionary.count == 1);
        XCTAssertEqualObjects(dictionary[@1], user.objectID);

        NSManagedObjectID *objectID = dictionary[@1];
        User *retreivedUser = (User *)[context objectWithID:objectID];
        XCTAssertEqualObjects(retreivedUser.remoteID, @1);
        XCTAssertEqualObjects(retreivedUser.name, @"Joshua");
    }];
}

- (void)testDictionaryStringLocalKey {
    [self configureUserWithRemoteID:nil localID:@"100" name:@"Joshua" block:^(User *user, NSManagedObjectContext *context) {
        NSDictionary *dictionary = [DATAObjectIDs objectIDsInEntityNamed:@"User"
                                                     withAttributesNamed:@"localID"
                                                                 context:context];
        XCTAssertNotNil(dictionary);
        XCTAssertTrue(dictionary.count == 1);
        XCTAssertEqualObjects(dictionary[@"100"], user.objectID);

        NSManagedObjectID *objectID = dictionary[@"100"];
        User *retreivedUser = (User *)[context objectWithID:objectID];
        XCTAssertEqualObjects(retreivedUser.localID, @"100");
        XCTAssertEqualObjects(retreivedUser.name, @"Joshua");
    }];
}

- (void)testObjectIDsArray {
    [self configureUserWithRemoteID:@1 localID:nil name:@"Joshua" block:^(User *user, NSManagedObjectContext *context) {
        NSArray *objectIDs = [DATAObjectIDs objectIDsInEntityNamed:@"User"
                                                           context:context];
        XCTAssertNotNil(objectIDs);
        XCTAssertEqual(objectIDs.count, 1);
        XCTAssertEqualObjects(objectIDs.firstObject, user.objectID);
    }];
}

- (void)testObjectIDsArrayWithPredicate {
    DATAStack *stack = [[DATAStack alloc] initWithModelName:@"Tests" bundle:[NSBundle bundleForClass:[self class]]
                                                  storeType:DATAStackStoreTypeInMemory];

    [self insertUserWithRemoteID:@1 localID:nil name:@"Joshua" inContext:stack.mainContext];
    User *jon = [self insertUserWithRemoteID:@2 localID:nil name:@"Jon" inContext:stack.mainContext];

    NSPredicate *predicate = [NSPredicate predicateWithFormat:@"name == 'Jon'"];
    NSArray *objectIDs = [DATAObjectIDs objectIDsInEntityNamed:@"User"
                                                       context:stack.mainContext
                                                     predicate:predicate];
    XCTAssertNotNil(objectIDs);
    XCTAssertEqual(objectIDs.count, 1);
    XCTAssertEqualObjects(objectIDs.firstObject, jon.objectID);
}

- (void)testDictionaryStringLocalKeyUsingSortDescriptor {
    DATAStack *stack = [[DATAStack alloc] initWithModelName:@"Tests" bundle:[NSBundle bundleForClass:[self class]]
                                                  storeType:DATAStackStoreTypeInMemory];
    [stack performInNewBackgroundContext:^(NSManagedObjectContext *context) {
        [self insertUserWithRemoteID:nil localID:@"100" name:@"Joshua" inContext:context];
        [self insertUserWithRemoteID:nil localID:@"200" name:@"Jon" inContext:context];
        [context save:nil];

        NSArray *attributesA = [DATAObjectIDs attributesInEntityNamed:@"User"
                                                        attributeName:@"localID"
                                                              context:context
                                                      sortDescriptors:@[[NSSortDescriptor sortDescriptorWithKey:@"localID" ascending:YES]]];
        XCTAssertEqualObjects(attributesA.firstObject, @"100");

        NSArray *attributesB = [DATAObjectIDs attributesInEntityNamed:@"User"
                                                        attributeName:@"localID"
                                                              context:context
                                                      sortDescriptors:@[[NSSortDescriptor sortDescriptorWithKey:@"localID" ascending:NO]]];
        XCTAssertEqualObjects(attributesB.firstObject, @"200");
    }];
}

@end

File: Tests/DATAObjectIDs/DATAObjectIDsTests.swift
import XCTest
import DATAStack

class DATAObjectIDsTests: XCTestCase {
    func insertUserWithRemoteID(remoteID: NSNumber?, localID: String?, name: String, context: NSManagedObjectContext) -> User {
        let user = NSEntityDescription.insertNewObject(forEntityName: "User", into: context) as! User
        user.remoteID = remoteID
        user.localID = localID
        user.name = name
        
        return user
    }

    func configureUserWithRemoteID(remoteID: NSNumber?, localID: String?, name: String, block: @escaping (_ user: User, _ context: NSManagedObjectContext) -> Void) {
        let stack = DATAStack(modelName: "Tests", bundle: Bundle(for: DATAObjectIDsTests.self), storeType: .inMemory)
        stack.performInNewBackgroundContext { context in
            let user = self.insertUserWithRemoteID(remoteID: remoteID, localID: localID, name: name, context: context)
            try! context.save()
            block(user, context)
        }
    }

    func testDictionary() {
        self.configureUserWithRemoteID(remoteID: 1, localID: nil, name: "Joshua") { user, context in
            let dictionary = DATAObjectIDs.objectIDs(inEntityNamed: "User", withAttributesNamed: "remoteID", context: context)
            XCTAssertNotNil(dictionary)
            XCTAssertTrue(dictionary.count == 1)
            XCTAssertEqual(dictionary[NSNumber(value: 1)], user.objectID)

            let objectID = dictionary[NSNumber(value: 1)]!
            let retreivedUser = context.object(with: objectID) as! User
            XCTAssertEqual(retreivedUser.remoteID, 1);
            XCTAssertEqual(retreivedUser.name, "Joshua");
        }
    }

    func testDictionaryStringLocalKey() {
        self.configureUserWithRemoteID(remoteID: nil, localID: "100", name: "Joshua") { user, context in
            let dictionary = DATAObjectIDs.objectIDs(inEntityNamed: "User", withAttributesNamed: "localID", context: context)
            XCTAssertNotNil(dictionary)
            XCTAssertTrue(dictionary.count == 1)
            XCTAssertEqual(dictionary["100"], user.objectID);

            let objectID = dictionary["100"]!
            let retreivedUser = context.object(with: objectID) as! User
            XCTAssertEqual(retreivedUser.localID, "100")
            XCTAssertEqual(retreivedUser.name, "Joshua")
        }
    }

    func testObjectIDsArray() {
        self.configureUserWithRemoteID(remoteID: 1, localID: nil, name: "Joshua") { user, context in
            let objectIDs = DATAObjectIDs.objectIDs(inEntityNamed: "User", context: context)
            XCTAssertEqual(objectIDs.count, 1);
            XCTAssertEqual(objectIDs.first, user.objectID)
        }
    }

    func testObjectIDsArrayWithPredicate() {
        let stack = DATAStack(modelName: "Tests", bundle: Bundle(for: DATAObjectIDsTests.self), storeType: .inMemory)
        let _ = self.insertUserWithRemoteID(remoteID: 1, localID: nil, name: "Joshua", context: stack.mainContext)
        let jon = self.insertUserWithRemoteID(remoteID: 2, localID: nil, name: "Jon", context: stack.mainContext)

        let predicate = NSPredicate(format: "name == 'Jon'")
        let objectIDs = DATAObjectIDs.objectIDs(inEntityNamed: "User", context: stack.mainContext, predicate: predicate) 
        XCTAssertEqual(objectIDs.count, 1)
        XCTAssertEqual(objectIDs.first, jon.objectID)
    }

    func testDictionaryStringLocalKeyUsingSortDescriptor() {
        let stack = DATAStack(modelName: "Tests", bundle: Bundle(for: DATAObjectIDsTests.self), storeType: .inMemory)
        stack.performInNewBackgroundContext { context in
            let _ = self.insertUserWithRemoteID(remoteID: nil, localID: "100", name: "Joshua", context: context)
            let _ = self.insertUserWithRemoteID(remoteID: nil, localID: "200", name: "Jon", context: context)
            try! context.save()

            let attributesA = DATAObjectIDs.attributes(inEntityNamed: "User", attributeName: "localID", context: context, sortDescriptors: [NSSortDescriptor(key: "localID", ascending: true)])
            XCTAssertEqual(attributesA.first as? String, "100")

            let attributesB = DATAObjectIDs.attributes(inEntityNamed: "User", attributeName: "localID", context: context, sortDescriptors: [NSSortDescriptor(key: "localID", ascending: false)])
            XCTAssertEqual(attributesB.first as? String, "200")
        }
    }
}
